src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/Macros/GeneratePortNumberConfig.cs

```diff
@@ -0,0 +1,145 @@
+        {
+            int low = GetOptionalParameterValue(generatedSymbolConfig, nameof(Low), ConvertJTokenToInt, LowPortDefault);
+            int high = GetOptionalParameterValue(generatedSymbolConfig, nameof(High), ConvertJTokenToInt, HighPortDefault);
+            if (low < LowPortDefault)
+            {
+                low = LowPortDefault;
+            }
+
+            if (high > HighPortDefault)
+            {
+                high = HighPortDefault;
+            }
+
+            if (low > high)
+            {
+                low = LowPortDefault;
+                high = HighPortDefault;
+            }
+
+            int fallback = GetOptionalParameterValue(generatedSymbolConfig, "fallback", ConvertJTokenToInt, 0);
+
+            Port = AllocatePort(low, high, fallback);
+        }
+
+        internal int Port { get; }
+
+        internal int Low { get; }
+
+        internal int High { get; }
+
+        private static int AllocatePort(int low, int high, int fallback = 0)
+        {
+            int startPort = CryptoRandom.NextInt(low, high);
+
+            for (int testPort = startPort; testPort <= high; testPort++)
+            {
+                if (TryAllocatePort(testPort, out int port))
+                {
+                    return port;
+                }
+            }
+
+            for (int testPort = low; testPort < startPort; testPort++)
+            {
+                if (TryAllocatePort(testPort, out int port))
+                {
+                    return port;
+                }
+            }
+            return fallback;
+        }
+
+        private static bool TryAllocatePort(int testPort, out int finalPort)
+        {
+            Socket? testSocket = null;
+            finalPort = 0;
+
+            if (UnsafePorts.Contains(testPort))
+            {
+                return false;
+            }
+            lock (LockObj)
+            {
+                if (AllocatedPorts.Contains(testPort))
+                {
+                    return false;
+                }
+
+                try
+                {
+                    if (Socket.OSSupportsIPv4)
+                    {
+                        testSocket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+                    }
+                    else if (Socket.OSSupportsIPv6)
+                    {
+                        testSocket = new Socket(AddressFamily.InterNetworkV6, SocketType.Stream, ProtocolType.Tcp);
+                    }
+
+                    if (testSocket is null)
+                    {
+                        return false;
+                    }
+                    IPEndPoint endPoint = new(testSocket.AddressFamily == AddressFamily.InterNetworkV6 ? IPAddress.IPv6Any : IPAddress.Any, testPort);
+                    testSocket.Bind(endPoint);
+                    finalPort = ((IPEndPoint)testSocket.LocalEndPoint).Port;
+                    if (testPort != finalPort)
```

Comment:
This test is unnecessary.
The socket should not bind on a different port than it was configured to be bound (otherwise it would be a serious bug in networking primitives). But even if that would happen - we've found a free port - so it is a success